From 966a053e7689f41807d8994a93d5fd3952e09932 Mon Sep 17 00:00:00 2001
From: Marco Martin <mart@kde.org>
Date: Thu, 6 Aug 2015 14:03:59 +0200
Subject: [PATCH] Allow for styles to provide own EditMenu

Allow for styles to define their own EditMenu component, to provide
cut/copy/paste actions with the platform style.
In order to be able to provide mobile-optimized controls on Linux
systems, use the boolean env var QT_MOBILE_PLATFORM in this case.
This allows on a plain Linux system to offer controls with the
touch friendly text selection handles and the little inline
cut/copy/paste menu

Change-Id: Id9013b4386423938735390325a003fcccdcfd598
Reviewed-by: Richard Moe Gustavsen <richard.gustavsen@theqtcompany.com>
---
 src/controls/Private/EditMenu.qml           | 13 ++++++++++---
 src/controls/Styles/Base/TextAreaStyle.qml  |  6 ++++++
 src/controls/Styles/Base/TextFieldStyle.qml |  6 ++++++
 3 files changed, 22 insertions(+), 3 deletions(-)

diff --git a/src/controls/Private/EditMenu.qml b/src/controls/Private/EditMenu.qml
index 51abe75..dc3ba64 100644
--- a/src/controls/Private/EditMenu.qml
+++ b/src/controls/Private/EditMenu.qml
@@ -45,7 +45,7 @@ Loader {
     property Item selectionHandle
     property Flickable flickable
     property Component defaultMenu: item && item.defaultMenu ? item.defaultMenu : null
-    property Menu menuInstance: null
+    property QtObject menuInstance: null
     property MouseArea mouseArea
 
     Connections {
@@ -67,6 +67,13 @@ Loader {
         return menuInstance;
     }
 
-    source: Qt.resolvedUrl(Qt.platform.os === "ios" ? "EditMenu_ios.qml"
-                                                    : Qt.platform.os === "android" ? "" : "EditMenu_base.qml")
+    Component.onCompleted: {
+        if (__style.__editMenu) {
+            sourceComponent = __style.__editMenu;
+        } else {
+            // todo: get ios/android/base menus from style as well
+            source = (Qt.resolvedUrl(Qt.platform.os === "ios" ? "EditMenu_ios.qml"
+                : Qt.platform.os === "android" ? "" : "EditMenu_base.qml"));
+        }
+    }
 }
diff --git a/src/controls/Styles/Base/TextAreaStyle.qml b/src/controls/Styles/Base/TextAreaStyle.qml
index 678d365..1da5222 100644
--- a/src/controls/Styles/Base/TextAreaStyle.qml
+++ b/src/controls/Styles/Base/TextAreaStyle.qml
@@ -146,4 +146,10 @@ ScrollViewStyle {
         \since QtQuick.Controls.Styles 1.3
     */
     property Component __cursorDelegate
+
+    /*! \internal
+        The delegate for the cut/copy/paste menu.
+        \since QtQuick.Controls.Styles 1.4
+    */
+    property Component __editMenu
 }
diff --git a/src/controls/Styles/Base/TextFieldStyle.qml b/src/controls/Styles/Base/TextFieldStyle.qml
index e924741..b5e024b 100644
--- a/src/controls/Styles/Base/TextFieldStyle.qml
+++ b/src/controls/Styles/Base/TextFieldStyle.qml
@@ -209,4 +209,10 @@ Style {
         \since QtQuick.Controls.Styles 1.3
     */
     property Component __cursorDelegate
+
+    /*! \internal
+        The delegate for the cut/copy/paste menu.
+        \since QtQuick.Controls.Styles 1.4
+    */
+    property Component __editMenu
 }
-- 
2.6.2

